File: CP/LeetCode/Python/13.py
# 13. Roman to Integer

from __future__ import annotations


class Solution:
    def romanToInt(self, s: str) -> int:
        roman = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000,
        }
        return sum(
            roman[s[i]] - 2 * roman[s[i - 1]]
            if i > 0 and roman[s[i]] > roman[s[i - 1]]
            else roman[s[i]]
            for i in range(len(s))
        )


if __name__ == "__main__":
    s = Solution()
    print(s.romanToInt("III"))  # 3
    print(s.romanToInt("IV"))  # 4
    print(s.romanToInt("IX"))  # 9
    print(s.romanToInt("LVIII"))  # 58
    print(s.romanToInt("MCMXCIV"))  # 1994

File: CP/LeetCode/C/13.c
// 13. Roman to Integer

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int romanToInt(char *s) // NOLINT(readability-non-const-parameter)
{
    int i = 0;
    int num = 0;

    while (s[i] != '\0')
    {
        if (s[i] == 'I')
        {
            if (s[i + 1] == 'V')
            {
                num += 4;
                i += 2;
            }
            else if (s[i + 1] == 'X')
            {
                num += 9;
                i += 2;
            }
            else
            {
                num += 1;
                i++;
            }
        }
        else if (s[i] == 'V')
        {
            num += 5;
            i++;
        }
        else if (s[i] == 'X')
        {
            if (s[i + 1] == 'L')
            {
                num += 40;
                i += 2;
            }
            else if (s[i + 1] == 'C')
            {
                num += 90;
                i += 2;
            }
            else
            {
                num += 10;
                i++;
            }
        }
        else if (s[i] == 'L')
        {
            num += 50;
            i++;
        }
        else if (s[i] == 'C')
        {
            if (s[i + 1] == 'D')
            {
                num += 400;
                i += 2;
            }
            else if (s[i + 1] == 'M')
            {
                num += 900;
                i += 2;
            }
            else
            {
                num += 100;
                i++;
            }
        }
        else if (s[i] == 'D')
        {
            num += 500;
            i++;
        }
        else if (s[i] == 'M')
        {
            num += 1000;
            i++;
        }
    }

    return num;
}

int main()
{
    char s[100] = "MCMXCIV";
    printf("%d\n", romanToInt(s));

    return 0;
}